cfgov/v1/tests/test_util.py

```diff
@@ -19,57 +23,61 @@ def setUp(self):
-            mock.Mock(**{'content_json': {'live': False, 'shared': False}}),
-            mock.Mock(**{'content_json': {'live': False, 'shared': False}}),
-            mock.Mock(**{'content_json': {'live': False, 'shared': False}}),
-        ]
-        mock_json_loads.side_effect = [page.content_json for page in self.page_versions]
-        util.get_appropriate_page_version(self.request, self.page)
-        assert not self.page_versions[0].as_page_object.called
-        assert not self.page_versions[1].as_page_object.called
-        assert not self.page_versions[2].as_page_object.called
-        assert not self.page_versions[3].as_page_object.called
+        self.new_page = CFGOVPage(title='a cfgov page')
+        self.new_page.live = False
+        self.new_page.shared = False
 
+        helpers.save_new_page(self.new_page)
+        content_json = json.loads(self.new_page.to_json())
+        # create the various page revisions
+        # revision 1
+        content_json['title'] = 'revision 1'
+        content_json['live'] = True
+        content_json['shared'] = True
+        self.revision_1 = self.new_page.revisions.create(content_json=json.dumps(content_json))
 
-    @mock.patch('json.loads')
-    def test_shared_returns_None_if_page_not_live_when_on_production(self, mock_json_loads):
-        self.request.site.hostname = 'localhost'
-        self.page_versions = [
-            mock.Mock(**{'content_json': {'live': False, 'shared': True}}),
-            mock.Mock(**{'content_json': {'live': False, 'shared': True}}),
-            mock.Mock(**{'content_json': {'live': False, 'shared': True}}),
-            mock.Mock(**{'content_json': {'live': False, 'shared': True}}),
-        ]
-        mock_json_loads.side_effect = [page.content_json for page in self.page_versions]
-        util.get_appropriate_page_version(self.request, self.page)
-        assert not self.page_versions[0].as_page_object.called
-        assert not self.page_versions[1].as_page_object.called
-        assert not self.page_versions[2].as_page_object.called
-        assert not self.page_versions[3].as_page_object.called
+        # rev 2
+        content_json['title'] = 'revision 2'
+        content_json['live'] = True
+        content_json['shared'] = False
+        self.revision_2 = self.new_page.revisions.create(content_json=json.dumps(content_json))
+
+        # rev 3
+        content_json['title'] = 'revision 3'
+        content_json['live'] = False
+        content_json['shared'] = True
+        self.revision_3 = self.new_page.revisions.create(content_json=json.dumps(content_json))
+
+        # rev 4
+        content_json['title'] = 'revision 4'
+        content_json['live'] = False
+        content_json['shared'] = False
+        self.revision_4 = self.new_page.revisions.create(content_json=json.dumps(content_json))
+
+    def test_production_returns_first_live_page(self):
+        self.request.is_staging = False
```

Comment:
> One comment on "they are carried over from test to test" -- there's no guarantee of test order so changes in one test to a class property shouldn't be required for other tests to pass. Each test should only count on the initialization in setUp; there shouldn't be anything else getting carried over.

True, I hope I didn't imply that this was the case but if I did I just want to make it clear that I recognize this.

>Let's say this test fails because of some changes and I'm looking at it as someone new to this code. It starts with deletion of three class member variables, which I have no way of understanding without looking through setUp. It calls the method under test using another class member self.request which isn't defined locally. I really have to dig into to all of these variables to understand what this is supposed to be testing.

Yeah, this isn't great; I could probably change this to just pull up the various revisions and change their `live` and `shared` properties to match the test conditions. I think for a test with more complicated logic your approach would be the more correct one.